srk/src/chc.ml

```diff
@@ -120,120 +176,312 @@ module Fp = struct
+    BatArray.iteri (fun scc1 rel1 ->
+        BatArray.iteri (fun scc2 rel2 ->
+            if D.PathCheck.check_path path_check rel1 rel2 then
+              D.add_edge sccgraph (sccverts.(scc1)) (sccverts.(scc2))
+            else ())
+          scc_rep)
+      scc_rep;
+    let solved = Hashtbl.create (num_rels + 1) in
+    (* We compute a topologoical sort on the collapsed graph determine
+     * if the collapsed graph forms a stratified lin system of chcs.*)
+    let (ordering, is_strat) = 
+      D.Topological.fold (fun scc_num (ordering, is_stratifiable) ->
+          (* First we extract the sub-chc *)
+          let of_vert arr v = BatArray.findi ((=) v) arr in
+          let rels = 
+            Relation.Set.of_list 
+              (List.map (of_vert verts) sccs.(of_vert sccverts scc_num)) 
+          in
+          let ordering' = rels :: ordering in
+          let ruleset = 
+            List.filter (fun (_, _, conc_atom) ->
+                Relation.Set.mem (rel_of_atom conc_atom) rels)
+              fp.rules
+          in
+          let is_stratifiable' = List.fold_left (fun acc (hypo_atoms, _, _) ->
+              if List.length (List.filter (fun atom ->
+                  not (Hashtbl.mem solved (rel_of_atom atom)))
+                  hypo_atoms) > 1
+              then false
+              else true && acc)
+              is_stratifiable
+              ruleset
+          in
+          Relation.Set.iter 
+            (fun rel -> Hashtbl.add solved rel ()) 
+            rels;
+          (ordering', is_stratifiable')
+        )
+        sccgraph
+        ([], true)
+    in
+    if is_strat then Some (List.rev ordering) else None
+
+
+  let solve_super_lin srk fp pd ordering =
+    let fp = normalize srk fp in
+    let num_rels = (DynArray.length fp.rel_ctx) - 1 in
+    let solution = Hashtbl.create (num_rels + 1) in
+    (* We compute a topologoical sort on the collapsed graph and solve
+     * for the relations in order.*)
+    List.iter (fun rels ->
+        let ruleset = 
+          List.filter (fun (_, _, conc_atom) ->
+              Relation.Set.mem (rel_of_atom conc_atom) rels)
+            fp.rules
+        in
+        (* Substitute in the solutions for rels calculated at previous strata *)
+        let ruleset' = List.map (fun (hypo_atoms, constr, conc) ->
+            let hypo_atoms', constr' = 
+              List.fold_left (fun (hypo_atoms, constr) atom -> 
+                  if Hashtbl.mem solution (rel_of_atom atom) then (
+                    let syms, soln = Hashtbl.find solution (rel_of_atom atom) in
+                    let subst =
+                      (fun sym ->
+                         mk_const
+                           srk
+                           (List.nth
+                              (params_of_atom atom)
+                              (BatArray.findi ((=) sym) syms)))
+                    in
+                    let soln = substitute_const srk subst soln in
```

Comment:
Is the solution guaranteed to only use the symbols in syms?

Reply:
If my code is correct, that should be the case. "syms, soln" is the solution to a relation symbol where syms[n] is the symbol used for nth argument to that relation. Solve_linear_chc converts the chc to a graph where the nodes are relations and the edges are constraints whose free variables are limited to the symbols used as arguments to the input and output relation. This invariant is initially ensured by calling the normalization procedure and then is maintained through the add, mul, and star function of the algebra.